### PR TITLE
M-FSDP: Make fine_grained_param_gather configurable for MXFP8 to enable performance–memory trade-offs

### PR DESCRIPTION
**Pull request by shjwudp**
# What does this PR do ?

This change updates the MXFP8 implementation so that `fine_grained_param_gather`, previously always enabled, is now exposed as a configurable option.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 1]
```diff
@@ -265,7 +264,9 @@ def __init__(
         self.mp_policy = mixed_precision_policy
         self.calculate_per_token_loss = calculate_per_token_loss
         self.init_model_with_meta_device = init_model_with_meta_device
-        self.enable_fine_grained_param_gather_hook = enable_fine_grained_param_gather_hook
+        self.enable_fine_grained_param_gather_hook = (
```
Comment (sanandaraj5597) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: @shjwudp Can we override this to be True when we have fp8_recipe as MX-FP8 and fp8_param_gather is True? Users might forget to set this for MX-FP8 and might face issues, so better to override right?

**Review by cspades — APPROVED** [review]
Definitely could use more documentation as well, it's not obvious how fine-grained / non-recursive module AG is related to row-wise / col-wise parameters.

The Megatron-FSDP logic that figures out what parameters to gather, as well as the TransformerEngine logic for when row-wise or col-wise data exists, should be explained in the future.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 1 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Isn't that what they were already doing? See `megatron/core/distributed/fsdp/mcore_fsdp_adapter.py`.

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Yes, but I think @shjwudp interpreted your feedback as "let's just make it a manual toggle".

What you want is OR logic:
```
enable_fine_grained_param_gather_hook=config.megatron_fsdp_enable_fine_grained_param_gather or (
    config.fp8_recipe == "mxfp8" and ddp_config.fp8_param_gather
),
```
Also nit: can we make this argument and config name shorter: `megatron_fsdp_enable_fine_grained_param_gather` -> `megatron_fsdp_fine_grained_param_ag`

Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: In that case, when we run any [mxfp8+native fp8 model param] config without recompute, we'll always run with fine-grained AG. I thought that's what we don't want.

I thought the idea is that we don't want fine-grained AG, especially when we don't have recompute.

Comment (sanandaraj5597) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: I think Ritesh is correct, lets not override it, but we should set it outside in M-LM training loop based on fp8_param, fp8_recipe and recompute modules present or not.

Comment (cspades) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Oh shoot, that is a typo, I forgot to add the recompute part: `args.recompute_granularity == 'selective'`.

Then to summarize, if we are using MXFP8 and are using activation recompute, then fine-grained AG must be used. Otherwise, you can either toggle it, or just not set it at all.

```
# Optional: OR set it as an argument.
enable_fine_grained_param_gather_hook=(
    config.fp8_recipe == "mxfp8" and ddp_config.fp8_param_gather and args.recompute_granularity == 'selective'
)
```

Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py:267]: This looks good to me. I'm not sure if there's a way to add what @cspades mentioned within MCore itself. But if there is, then it makes sense to have that as a heuristic.

**Comment by sanandaraj5597** [comment]
LGTM.

**Review by sanandaraj5597 — APPROVED** [review]

**Review by deepakn94 — APPROVED** [review]

**Review by ericharper — APPROVED** [review]
